trace/collector.go

```diff
@@ -0,0 +1,49 @@
+package trace
+
+import (
+	"log"
+	"sync"
+)
+
+type Collector interface {
+	Record(e Event) error
+}
+
+type MemoryCollector struct {
+	list []Event
+	in   chan Event
+	lock sync.Mutex
+}
+
+func (c MemoryCollector) Record(event Event) error {
+	c.in <- event
+	return nil
+}
+
+func (c MemoryCollector) GetEvents() []Event {
```

Comment:
I'm not 100% sure but I think if you are not using a pointer receiver then the `sync.Mutex` will be copied each time and on every method call you end up with a new mutex and locking will be non-effective. This is a problem because the `list` field is shared across all copies (as it is a slice).